src/auth/downscopedclient.ts

```diff
@@ -0,0 +1,254 @@
+}
+
+export class DownscopedClient extends AuthClient {
+  /**
+   * OAuth scopes for the GCP access token to use. When not provided,
+   * the default https://www.googleapis.com/auth/cloud-platform is
+   * used.
+   */
+  private cachedDownscopedAccessToken: CredentialsWithResponse | null;
+  private readonly stsCredential: sts.StsCredentials;
+  public readonly authClient: AuthClient;
+  public readonly credentialAccessBoundary: CredentialAccessBoundary;
+  public readonly eagerRefreshThresholdMillis: number;
+  public readonly forceRefreshOnFailure: boolean;
+
+  constructor(
+    private client: AuthClient,
+    private cab: CredentialAccessBoundary,
+    additionalOptions?: RefreshOptions
+  ) {
+    super();
+
+    // Check a number of 1-10 access boundary rules are defined within credential access boundary.
+    if (cab.accessBoundary.accessBoundaryRules.length === 0) {
+      throw new Error('At least one access boundary rule needs to be defined.');
+    } else if (cab.accessBoundary.accessBoundaryRules.length > 10) {
+      throw new Error('Access boundary rule exceeds limit, max 10 allowed.');
+    }
+
+    // Check at least one permission should be defined in each access boundary rule.
+    for (const rule of cab.accessBoundary.accessBoundaryRules) {
+      if (rule.availablePermissions.length === 0) {
+        throw new Error(
+          'At least one permission should be defined in access boundary rules.'
+        );
+      }
+    }
+
+    this.stsCredential = new sts.StsCredentials(STS_ACCESS_TOKEN_URL);
+    // Default OAuth scope. This could be overridden via public property.
+    this.cachedDownscopedAccessToken = null;
+    this.credentialAccessBoundary = cab;
+    this.authClient = client;
+    // As threshold could be zero,
+    // eagerRefreshThresholdMillis || EXPIRATION_TIME_OFFSET will override the
+    // zero value.
+    if (typeof additionalOptions?.eagerRefreshThresholdMillis !== 'number') {
+      this.eagerRefreshThresholdMillis = EXPIRATION_TIME_OFFSET;
+    } else {
+      this.eagerRefreshThresholdMillis = additionalOptions!
+        .eagerRefreshThresholdMillis as number;
+    }
+    this.forceRefreshOnFailure = !!additionalOptions?.forceRefreshOnFailure;
+  }
+
+  /**
+   * Provides a mechanism to inject Downscoped access tokens directly.
+   * When the provided credential expires, a new credential, using the
+   * external account options are retrieved.
+   * Notice DownscopedClient is the broker class mainly used for generate
+   * downscoped access tokens, it is unlikely we call this function in real
+   * use case.
+   * We implement to make this a helper function for testing all cases in getAccessToken().
+   * @param credentials The Credentials object to set on the current client.
+   */
+  setCredentials(credentials: Credentials) {
+    super.setCredentials(credentials);
+    this.cachedDownscopedAccessToken = credentials;
+  }
+
+  async getAccessToken(): Promise<GetAccessTokenResponse> {
```

Comment:
I think there is a problem in just returning `GetAccessTokenResponse` since it doesn't expose the expiration time.
Can we return an extended interface of `GetAccessTokenResponse` with an `expirationTime` field? We can use epoch time in seconds. This is needed so the token broker can expose this to the token consumer. Otherwise the token consumer can't accurately get this.